autorest/utility.go

```diff
@@ -138,13 +138,38 @@ func MapToValues(m map[string]interface{}) url.Values {
 	return v
 }
 
+// AsStringSlice method converts interface{} to []string. This expects a
+//that the parameter passed to be a slice or array of a type that has the underlying
+//type a string.
+func AsStringSlice(s interface{}) ([]string, error) {
+	v := reflect.ValueOf(s)
+	if v.Kind() != reflect.Slice && v.Kind() != reflect.Array {
+		return nil, NewError("autorest", "AsStringSlice", "the value's type is not an array.")
+	}
+	stringSlice := make([]string, 0, v.Len())
+
+	for i := 0; i < v.Len(); i++ {
+		stringSlice = append(stringSlice, v.Index(i).String())
+	}
+	return stringSlice, nil
+}
+
 // String method converts interface v to string. If interface is a list, it
-// joins list elements using separator.
+// joins list elements using the separators.
 func String(v interface{}, sep ...string) string {
-	if len(sep) > 0 {
-		return ensureValueString(strings.Join(v.([]string), sep[0]))
+	if len(sep) == 0 {
+		return ensureValueString(v)
+	}
+	stringSlice, ok := v.([]string)
+	if ok == false {
+		var err error
+		stringSlice, err = AsStringSlice(v)
+		if err != nil {
+			panic(fmt.Sprintf("autorest: Couldn't convert value to a string %s.", err))
+		}
 	}
-	return ensureValueString(v)
+	stringValue := strings.Join(stringSlice, strings.Join(sep, ""))
```

Comment:
The original implementation used sep[0] so it seems to me this change will introduce new behavior.  Is that intentional?

Reply:
It is intentional yes. I would say it introduces proper behavior (i.e. fixes a bug). e.g. if someone passed previously multiple separators only the first one would be used when the intent would be clearly to use multiple separators.

Reply:
Are there any cases in go-autorest where more than one separator is specified?  I suspect the reason the method has this sig is to treat the sep param as optional, not because they want more than one separator.

Reply:
That might have been the intent. In our case, the only use by us is coming from generated code for parameters, namely here https://github.com/Azure/autorest.go/blob/395cbeb9664982d64b89581d9a4b10c7ba427a26/src/Extensions.cs#L226

But the issue we have I assume if it is used by client code ( since its publicly exposed). Im thinking as a client, if I see method that takes multiple parameters as seperator, do I want it to actually do that or get a comment saying that oh actually only the first one is used and the rest will be ignored.

Reply:
actually given that is a join and not a split, doesn't really make sense to have more, so will leave it like it is to avoid any unforeseen consequences.